### PR TITLE
Topic fix pkey repeats

### PR DESCRIPTION
**Pull request by telephon**
See #2814.

This adds a class `FuncStream1`. Alternatives are of course possible, but this seemed the best eventually, not least because this is a non-imperative formulation.

### CODE REVIEW

File: testsuite/classlibrary/TestStream.sc [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	test_funcStream {
+		var str, envir;
+
+		envir = Environment.new;
+		envir.use {
+			~count = 0;
+			str = FuncStream({ ~count = ~count + 1 }, { ~count = 0 });
+		};
+
+		this.assert(str.nextN(8).postln == (1..8), "FuncStream should call the nextFunc in original environment");
```
Comment (brianlheim) [inline testsuite/classlibrary/TestStream.sc]: Wouldn't `assertEquals`, without the `postln`, be a more canonical/readable way to do these?

**Comment by brianlheim** [comment]
Hmm.. would it be too much to ask to also write some documentation for `FuncStream` and `FuncStream1` as part of this PR? So that users can know why they're different classes?

**Comment by telephon** [comment]
if we are all fine with its existence, I'll document it of course.

File: testsuite/classlibrary/TestStream.sc (hunk 1 above)
Comment (telephon) [inline testsuite/classlibrary/TestStream.sc]: ah yes, forgot that it exists.

**Comment by brianlheim** [comment]
oh yeah, good point haha. I don't have a strong/informed opinion one way or the other, but it seems to make sense to me.

**Comment by jamshark70** [comment]
I guess I'm not really opposed to FuncStream1, not very strongly, but you could do this with less code duplication by `Pfuncn(...).asStream`.

Or, if you wanted to be more systematic about it:

- Pfunc().asStream --> FuncStream
- Pfuncn().asStream --> FuncStreamN

And then Pkey could follow suit, as its current implementation is really mostly a shortcut for `Pfunc { |ev| ev[\theKey] }`.

I'm never sure what's the threshold where the usefulness of a new class overrides the cost. If the method lookup table is a cross product of every class selector x every method selector x 2 (TheClass and Meta_TheClass), in my heavily quarked environment, that's adding 19,260 slots to the table.

```
(
var selectors = IdentitySet.new;

Class.allClasses.do { |class|
	class.methods.do { |method|
		selectors.add(method.name)
	}
};

selectors.size
)
-> 9630
```

FuncStream1 is, I think, a borderline case. In principle, and in the absence of memory or maintenance cost, of course it makes sense. But the gain is miniscule, because there's already a way (`Pfuncn`) to make a stream that evaluates a function *n* times. It smells a little like bloat to me, but I don't have a very strong objection.

**Comment by telephon** [comment]
Good point, James, that is always a reason to worry of course. We could even try to calculate how much electricity each class costs per year.

I would really like to gain a better feeling of this trade off.

The reason I decided against `Pfuncn(...).asStream` is that the latter creates a Routine that allocates a routine default stack, even if you actually just want to call a function a couple of times.

The alternative is to use a local variable outside the closure. This is ok, too. It seemed to me that `FuncStream1` (or `N`) would be more useful than `FuncStream`. So another option is to add a little load on `FuncStream` by adding a `repeats` variable (defaulting to `inf` or `nil`).

**Comment by telephon** [comment]
Just for comparison, the implementation of `Pkey` without a `FuncStream1` would be:

```
// access a key from the input event
Pkey : Pattern {
	var	<>key, <>repeats;
	*new { |key, repeats|
		^super.newCopyArgs(key, repeats)
	}
	storeArgs { ^[key] }
		// avoid creating a routine
	asStream {
		var	keystream = key.asStream;
		var count = 0;
		^FuncStream(
			if(repeats.isNil) {
				{ |inevent| inevent !? { inevent[keystream.next(inevent)] } }
			} {
				{ |inevent| 
					var res;
					if(inevent.notNil and: { count < repeats }) { res = inevent[keystream.next(inevent)] };
					count = count + 1;
					res
				}
			}
		);
	}
}
```
(untested)

**Comment by jamshark70** [comment]
> The reason I decided against Pfuncn(...).asStream is that the latter creates a Routine that allocates a routine default stack, even if you actually just want to call a function a couple of times.

On that note, there's this unintended bit of humor:

```
	// avoid creating a routine
	asStream {
		var	keystream = key.asStream;
```

An alternate approach would be to implement a proper `embedInStream`. I think there are just a couple of cases:

- Pkey with repeats is meant to control the total number of events output by the containing Pbind. In this case, a `FuncStream1` approach would avoid allocating a routine stack, while `embedInStream` would not.

- Pkey with repeats is embedded in a list pattern (Pseq, Prand etc.) -- in which case `embedInStream` would be called within the routine that already exists for the list pattern.

My gut feeling is that the first case would be relatively rare -- and, in that case, you would already be creating several other short-lived routines, so one more would not be so much less efficient. I guess the second case is likely to be more common.

Since #2814 isn't in response to a specific user case, we don't really know how this is going to be used.

> So another option is to add a little load on FuncStream by adding a repeats variable (defaulting to inf or nil).

That's reasonable.

**Comment by telephon** [comment]
>Since #2814 isn't in response to a specific user case, we don't really know how this is going to be used.

Ah I forgot to mention, I noticed this when trying:

`Pbind(\x, Pseq([1, 3, 2], inf), \y, Pseq([1, 2], inf), \note, Pseq([Pkey(\x, 1), Pkey(\y, 1)], inf))`

**Comment by jamshark70** [comment]
> Pseq([Pkey(\x, 1), Pkey(\y, 1)], inf)

In other words, Pkey instances embedded in a list pattern... which is the case I had guessed would be more likely. That one would avoid creating routines if Pkey implemented embedInStream.

**Comment by telephon** [comment]
yes. On the other hand, it might be useful to add a count to ` FuncStream` anyway.

**Comment by telephon** [comment]
Ok, here is a simpler implementation with a smaller footprint. There is a help file for `FuncStream` after all them years without.

**Comment by telephon** [comment]
When you merge, better squash.

**Comment by jamshark70** [comment]
If you want the list-pattern optimization, then Pkey must implement `embedInStream`.

I would do it this way. There's a bit of code duplication here, but I don't think it's that serious, and the result is:

- `repeats == inf` (normal case), no routine is created (it's a FuncStream instead).

- finite repeats, embedded in a list pattern, no routine is created.

- finite repeats, isolated, a routine is created (but this is a rare case).

```
// access a key from the input event
Pkey : Pattern {
	var	<>key, <>repeats;
	*new { |key, repeats|
		^super.newCopyArgs(key, repeats)
	}
	storeArgs { ^[key] }
	// for inf repeats, avoid creating a routine
	asStream {
		var	keystream;
		if(repeats.isNil or: { repeats == inf }) {
			keystream = key.asStream;
			^FuncStream({ |inevent|
				inevent !? { inevent[keystream.next(inevent)] }
			});
		} {
			^super.asStream
		}
	}
	embedInStream { |inval|
		var outval, keystream = key.asStream;
		(repeats.value(inval) ?? { inf }).do {
			outval = inval[keystream.next(inval)];
			if(outval.isNil) { ^inval };
			inval = outval.yield;
		};
		^inval
	}
}
```

```
Pseq([Pkey(\x, 1), Pkey(\y, 2)], inf).asStream.nextN(6, (x: 5, y: 500));
-> [ 5, 500, 500, 5, 500, 500 ]
```

**Comment by telephon** [comment]
Ah yes, I see. What about this simplification?

```
// access a key from the input event
Pkey : Pattern {
	var	<>key, <>repeats;
	*new { |key, repeats|
		^super.newCopyArgs(key, repeats ? inf)
	}
	storeArgs { ^[key, repeats] }
	// for inf repeats, avoid creating a routine
	asStream {
		var	keystream;
		^if(repeats == inf) {
			keystream = key.asStream;
			FuncStream({ |inevent|
				inevent !? { inevent[keystream.next(inevent)] }
			})
		} {
			super.asStream
		}
	}
	embedInStream { |inval|
		var outval, keystream = key.asStream;
		repeats.value(inval).do {
			outval = inval[keystream.next(inval)];
			if(outval.isNil) { ^inval };
			inval = outval.yield;
		};
		^inval
	}
}
```

**Comment by jamshark70** [comment]
OK, it's one fewer byte code, so I guess `^if...` is better.

I guess I wrote it the other way because I had just been working on C++ primitives, where it's common to have multiple `return` expressions in different branches.

**Comment by telephon** [comment]
I actually wrote it that way to convey that it will always return something else than `this`. But that's really more a matter of style - interesting that you can save bytecodes that way.

Note that in the version that is in 0a13e6c84, I kept the shorter variant, using `fin`. Let me know if you have reservations!

**Comment by jamshark70** [comment]
I think the `fin` version is OK. `fin` would be invoked mainly in cases like this: `Pbind(\dur, Pwhite(1, 4, inf) * 0.25, \sustain, Pkey(\dur, 4) + 1.5, ...)` and I think this is not likely to be common, so, no need to go out of the way to avoid a Routine here.

**Comment by jamshark70** [comment]
> I actually wrote it that way to convey that it will always return something else than this. But that's really more a matter of style - interesting that you can save bytecodes that way.

Well, if we want to be finicky, while I was investigating the weird stack trace that I posted to the mailing list yesterday, I found this in View:

```
	keyDown { arg char, modifiers, unicode, keycode, key;
		if( keyDownAction.notNil ) {
			^keyDownAction.value( this, char, modifiers, unicode, keycode, key );
		} {
			^this.defaultKeyDownAction( char, modifiers, unicode, keycode, key );
		};
	}
```

... where the multiple returns allow tail-call optimization to take effect. So, you win some, you lose some ;-p (but in this case, neither `FuncStream.new` nor `something.asStream` will have a deep stack, so it doesn't really matter -- but the stack depth for GUI response functions is impossible to predict because it's user code).

**Comment by telephon** [comment]
Very good point to remember. 

I wonder if the compiler could automatically optimize `^if(b) { x } { y }` to `if(b) { ^x } { ^y }` for tail call optimization.

**Comment by telephon** [comment]
This is ready to be merged.

**Comment by telephon** [comment]
these are good changes, thanks.

**Comment by brianlheim** [comment]
@telephon np! I actually just realized that we also now don't have documentation for the `repeats` arg for `Pkey.new`. I added some, if it looks good to you then please merge!

**Comment by jamshark70** [comment]
We missed something in this review.

```
(
Pdef(\moo, Pbind(
    \stuff, Pwhite(30,36,inf),
    \midinote, Pseq([Pkey(\stuff), 62, 64, 65], inf)
)).play;
)
```

In `embedInStream`, Pkey doesn't replace the default `repeats` = nil with any number, so it repeats 0 times.

Oh wait, I see I had proposed `(repeats.value(inval) ?? { inf }).do` for backward compatibility, but then the very next simplification discarded this :open_mouth: (which I then didn't notice :disappointed: ).

According to user reports, Pkey in Pseq used to act like `repeats == 1` (which I didn't expect).

Anyway, same old song and dance: *every time* you think, "Oh, this is just a simple change, it doesn't need a spec," you invite the risk of making mistakes.

I think we may have justification to start asking pull request authors to explain in text all of the expected cases, and reject the PR if they can't or don't.

**Comment by brianlheim** [comment]
> I think we may have justification to start asking pull request authors to explain in text all of the expected cases, and reject the PR if they can't or don't.

I think the current amount of churn is perfectly acceptable, so long as we introduce tests alongside the fixes. Then we can be sure that we don't regress on old bugs while fixing the new.

**Comment by telephon** [comment]
I am unsure in this case whether we need something like `streamArg`.

But I would assume that we simply do `(repeats.value(inval) ? 1).do {`

**Comment by jamshark70** [comment]
I think `Pkey(\something)` should have one and only one meaning: either an infinite stream in every case, or a single-value stream in every case. I think it's a hornet's nest if we try to say *sometimes* it's infinite-length and sometimes not -- unless there's a spec for it, but we're still not drawing up specs, hm. We're just going on a feeling that it might be nicer if Pkey limited its length automagically in Pseq.

Tbh I'm not clear on streamArg and I'm not at my computer now. That might be the way to go but I can't evaluate that just now.

**Comment by telephon** [comment]
the main difference is between embedding and streaming. You do want different behaviour there.

**Comment by jamshark70** [comment]
I don't necessarily disagree with that ("the main difference is between embedding and streaming. You do want different behaviour there") but to do this right needs a spec and a thorough review of patterns.

Here we have a case where backward compatibility was broken. I would argue that it's reasonable to restore backward compatibility before undertaking a larger change.

We have a history of starting broad-scope interface changes with a well-intentioned but largely unplanned localized change, and then getting backed into a corner later when trying to expand the scope of the new principle. Now in this case, it looks to me like you're arguing to repeat the mistakes of the past. I'm sorry to be obstinate, but I feel very strongly that planning is something that we, as a developer community, have never put much attention into -- and as a result, we don't do it well. Perhaps another result is that we don't think planning matters -- as here, where the proposal is to introduce a new principle almost casually.

Concretely, are we absolutely sure the difference between streaming and embedding is consistent throughout the class library? I have not much confidence in that.

**Comment by jamshark70** [comment]
PS I'm aware it's been discussed before, but the prior discussion didn't produce a spec.

**Pull request merged**